libs/ledger/src/storage_unit/transaction_store_sync_service.cpp

```diff
@@ -274,67 +312,77 @@ TransactionStoreSyncService::State TransactionStoreSyncService::OnResolvingSubtr
   {
     FETCH_LOG_WARN(LOGGING_NAME, "Lane ", cfg_.lane_id, ": ", "Failed subtree promises count ",
                    counts.failed);
+
     for (auto &fail : pending_subtree_.GetFailures(MAX_SUBTREE_RESOLUTION_PER_CYCLE))
     {
       roots_to_sync_.push(promise_id_to_roots_[fail.promise.id()]);
     }
+
+    subtree_failure_total_->add(static_cast<uint64_t>(counts.failed));
   }
-  if (counts.pending > 0)
+
+  // evaluate if the syncing process if complete, this can only be the case when there are no in
+  // flight requests and we have successfully evaluated all the roots we are after
+  bool const is_subtree_sync_complete = roots_to_sync_.empty() && counts.pending == 0;
+
+  if (!is_subtree_sync_complete)
   {
-    if (!promise_wait_timeout_.IsDue())
-    {
-      if (!roots_to_sync_.empty())
-      {
-        return State::QUERY_SUBTREE;
-      }
+    // determine what is the max number of in flight promises we should issue
+    auto const        directly_connected_peers = muddle_->AsEndpoint().GetDirectlyConnectedPeers();
+    std::size_t const maximum_inflight = MAX_REQUESTS_PER_NODE * directly_connected_peers.size();
 
-      return State::RESOLVING_SUBTREE;
+    bool const should_schedule_next_roots =
+        (counts.pending < maximum_inflight) && (!roots_to_sync_.empty());
+
+    // if there are still roots to be synced and we have not reached our maximum number of inflight
+    // requests then progress to the
+    if (should_schedule_next_roots)
+    {
+      return State::QUERY_SUBTREE;
     }
     else
     {
-      FETCH_LOG_WARN(LOGGING_NAME, "Lane ", cfg_.lane_id, ": ",
-                     "Timeout for subtree promises count!", counts.pending);
-      // get the pending
-      auto pending = pending_subtree_.GetPending();
-      for (auto &req : pending)
-      {
-        roots_to_sync_.push(promise_id_to_roots_[req.second.id()]);
-      }
+      // we are waiting for the in flight sub tree requests to terminate
+      state_machine_->Delay(10ms);
+      return State::RESOLVING_SUBTREE;
     }
   }
```

Comment:
I think it would be better to avoid duplication of the logic and simply do:
```cpp
bool const is_subtree_sync_complete = roots_to_sync_.empty() && counts.pending == 0
if (!is_subtree_sync_complete)
{
  state_machine_->Delay(10ms);
  return State::QUERY_SUBTREE;
}
```

That will simply execute `OnQuerySubtree()` where all of the logic is already implemented.